gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2

```diff
@@ -316,36 +322,110 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 client_cert_source = mtls.default_client_cert_source()
         return client_cert_source
 
-    def _get_api_endpoint(api_override, client_cert_source, use_mtls_endpoint):
+    def _get_api_endpoint(api_override, client_cert_source, universe_domain, use_mtls_endpoint):
         """Return the API endpoint used by the client.
 
         Args:
-            api_override (str): The API endpoint override. If specified, this is always the return value of this function.
+            api_override (str): The API endpoint override. If specified, this is always
+                the return value of this function and the other arguments are not used.
             client_cert_source (bytes): The client certificate source used by the client.
-            use_mtls_endpoint (str): How to use the MTLS endpoint, which depends also on the other parameters.
+            universe_domain (str): The universe domain used by the client.
+            use_mtls_endpoint (str): How to use the mTLS endpoint, which depends also on the other parameters.
                 Possible values are "always", "auto", or "never".
 
         Returns:
             str: The API endpoint to be used by the client.
         """
-
         if api_override is not None:
             api_endpoint = api_override
         elif use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source):
+            _default_universe = {{ service.client_name }}._DEFAULT_UNIVERSE
+            if universe_domain != _default_universe:
+                raise MutualTLSChannelError(f"mTLS is not supported in any universe other than {_default_universe}.")
             api_endpoint = {{ service.client_name }}.DEFAULT_MTLS_ENDPOINT
         else:
-            api_endpoint = {{ service.client_name }}.DEFAULT_ENDPOINT
+            api_endpoint = {{ service.client_name }}._DEFAULT_ENDPOINT_TEMPLATE.format(UNIVERSE_DOMAIN=universe_domain)
         return api_endpoint
 
+    @staticmethod
+    def _get_universe_domain(client_universe_domain, universe_domain_env):
```

Comment:
Add type hints for `client_universe_domain` and `client_universe_domain` as well as type hints for the return value.